libs/@local/harpc/service/src/session/server/write.rs

```diff
@@ -0,0 +1,263 @@
+use bytes::{Buf, Bytes};
+use bytes_utils::SegmentedBuf;
+use harpc_wire_protocol::{
+    flags::BitFlagsOp,
+    payload::Payload,
+    protocol::{Protocol, ProtocolVersion},
+    request::id::RequestId,
+    response::{
+        begin::ResponseBegin,
+        body::ResponseBody,
+        flags::{ResponseFlag, ResponseFlags},
+        frame::ResponseFrame,
+        header::ResponseHeader,
+        kind::{ErrorCode, ResponseKind},
+        Response,
+    },
+};
+use tokio::sync::mpsc;
+
+pub(crate) struct ResponseWriter<'a> {
+    id: RequestId,
+    kind: ResponseKind,
+    index: usize,
+
+    buffer: SegmentedBuf<Bytes>,
+
+    tx: &'a mpsc::Sender<Response>,
+
+    no_delay: bool,
+}
+
+impl<'a> ResponseWriter<'a> {
+    pub(crate) fn new_ok(id: RequestId, tx: &'a mpsc::Sender<Response>) -> Self {
+        Self {
+            id,
+            kind: ResponseKind::Ok,
+            index: 0,
+            buffer: SegmentedBuf::new(),
+            no_delay: false,
+
+            tx,
+        }
+    }
+
+    pub(crate) fn new_error(
+        id: RequestId,
+        error: ErrorCode,
+        tx: &'a mpsc::Sender<Response>,
+    ) -> Self {
+        Self {
+            id,
+            kind: ResponseKind::Err(error),
+            index: 0,
+            buffer: SegmentedBuf::new(),
+            no_delay: false,
+
+            tx,
+        }
+    }
+
+    pub(crate) const fn is_error(&self) -> bool {
+        matches!(self.kind, ResponseKind::Err(_))
+    }
+
+    pub(crate) fn with_no_delay(self, no_delay: bool) -> Self {
```

Comment:
`without_delay`?

Reply:
This is intentional, because it is `no_delay`, I specifically chose this name because of `TCP_NODELAY`, I'll write comments, what the setting exactly means.